src/main/java/at/hannibal2/skyhanni/features/inventory/chocolatefactory/RabbitUtils.kt

```diff
@@ -0,0 +1,70 @@
+package at.hannibal2.skyhanni.features.inventory.chocolatefactory
+
+import at.hannibal2.skyhanni.utils.CollectionUtils.getOrNull
+import kotlin.math.floor
+import kotlin.math.pow
+import kotlin.math.round
+
+object RabbitUtils {
+    /** Calculates the upgrade Cost for the upgrade at [slotIndex] with [level].
+     *
+     * @param slotIndex the slot index of the upgrade
+     * @param level the level of the upgrade
+     *
+     * @return the cost of the upgrade || null if the upgrade is maxed.
+     */
+    fun getUpgradeCost(slotIndex: Int, level: Int): Long? {
+        var price: Long? = null
+        if (level < (ChocolateFactoryAPI.maxUpgradeLevelPerPrestige[slotIndex]?.getOrNull(
+                ChocolateFactoryAPI.currentPrestige - 1,
+            ) ?: 0)
+        ) {
+
+            val prestigeMultiplier = 1 + (ChocolateFactoryAPI.upgradeCostFormulaConstants[slotIndex]?.get("prestige") ?: 0.0) *
+                (ChocolateFactoryAPI.currentPrestige - 1)
+            val zetaMultiplier = if (ChocolateFactoryAPI.foundZetaRabbit) 0.99 else 1.0
+
+            // Use upgrade cost per level if it exists, otherwise use the formula.
+            if ((ChocolateFactoryAPI.upgradeCostPerLevel[slotIndex]?.size ?: 0) > level) {
+                val nextRaw = ChocolateFactoryAPI.upgradeCostPerLevel[slotIndex]?.get(level) ?: 0
+
+                price = floor(nextRaw * prestigeMultiplier * zetaMultiplier).toLong()
+            } else {
+                val base = ChocolateFactoryAPI.upgradeCostFormulaConstants[slotIndex]?.get("base") ?: 0.0
+                val multiplier = ChocolateFactoryAPI.upgradeCostFormulaConstants[slotIndex]?.get("exp") ?: 0.0
+
+                price = floor(round(base * multiplier.pow((level + 1)) * prestigeMultiplier) * zetaMultiplier).toLong()
```

Comment:
I know I'm not a contributor here, but this should be 
``` 
price = floor(floor(round( base * multiplier.pow(level + 1)) * prestigeMultiplier) * zetaMultiplier).toLong()
```
to prevent quite a few off-by-one errors in prestiges that dont already have round prestigeMultipliers, such as the cost of level 12 Rabbit Sis at Chocolate Factory 4 with zeta rabbit, which is calculated as 12 069 by this formula but is actually 12 068.

Reply:
Thank you for taking your time reviewing the formulae. However the change you requested should not make a difference since x.toLong() will yield the same number as floor(x) in this case. But you made me recognize that there's an opening bracket to much in `multiplier.pow((level + 1))` that should instead be at the beginning.

Reply:
That's on me for incorrectly assuming the role of `.toLong()` in that line, otherwise the math checks out with your latest changes